### ISSUE TITLE
Allow `on` handlers to be asynchronous

### ISSUE THREAD

**Issue by RahulLanjewar93** [Secreto31126/whatsapp-api-js#195]
Currently, the return type for functions on the `on` handler is `void`. 
But realistically, it should support both synchronous as well as asynchronous callbacks.
https://github.com/Secreto31126/whatsapp-api-js/blob/4e4e68c755f900c5bd60458589c3825e30fb6021/src/emitters.d.ts#L18
https://github.com/Secreto31126/whatsapp-api-js/blob/4e4e68c755f900c5bd60458589c3825e30fb6021/src/emitters.d.ts#L60
https://github.com/Secreto31126/whatsapp-api-js/blob/4e4e68c755f900c5bd60458589c3825e30fb6021/src/emitters.d.ts#L94

Also we might need to handle these callbacks probably with a `try..catch`, if we decide to make it async 

https://github.com/Secreto31126/whatsapp-api-js/blob/4e4e68c755f900c5bd60458589c3825e30fb6021/src/index.ts#L780

I can create a PR for the same.
Thanks!

**Comment by Secreto31126** [comment]
I totally agree with you. The return type of the emitters should be `unknown`, hence supporting both sync and async callbacks.

However, I disagree about handling the return values in index. What's more, I think the methods should be called async in order to not block the main thread and also avoid errors (post() should NOT be affected by the emitters).

Maybe adding a new private method called `promisify` in the class could be usefull for this scenarios.

```typescript
// Quick draft
private async function promisify<T extends (...a: unknown[]) => unknown>(f: T, ...a: Params<T>) {
    f(...a);
}
```

<details>
  <summary><h3>Why not blocking post() thread</h3></summary>

  When the server receives a message, the expected response structure should be:

  `POST -> ACK (200) -> handler (most probably with a sendMessage)`

  However, if we await the return value of the event handlers, the graph would look like:

  `POST -> handler -> ACK (200)`

  This resolution, although not incorrect from the API perspective, could bring severe issues to post(), as the acknowledge should be sent in n seconds until Meta considers it failed. If handler is an expensive operation, post() might fail to ACK, causing Meta to resend the request to the server, which will call handler and fail to ACK because handler takes too long...

  TL;DR: "Infinite" loop.
</details>

If you want, feel free to open a PR to update the return types of the emitters, but don't fix index.ts on the same branch. I think it qualifies as a breaking change and can perfectly fit in v2.

Thanks!

**Comment by RahulLanjewar93** [comment]
Sure, I agree with the resolution structure you mentioned above.
I'll create a PR to fix the return types of emitters. 

But why do we need to set it as `unknown` shouldn't it be `void` or `Promise<void>`.
We don't expect or care about the return value in any case.

**Comment by RahulLanjewar93** [comment]
I think we can still achieve the desired resolution structure, without changing much

```typescript
    // Promise.resolve() will offload it from the main thread
    Promise.resolve()
              .then(() => {
                    // Add a try catch handler so that, if the callback is asynchronous, it will be handled 
                    try {
                        // Run the callback defined by the user.
                        this.on.message?.(args);
                    } catch (error) {
                        // Log the error, or do something else
                        console.warn(error);
                    }
                });
```

**Comment by Secreto31126** [comment]
Not expecting or caring for a return value doesn't mean it can't have one.

Imaging a scenario where the user has to use the following function to log to the terminal, either because the enviroment requires so or has a modified implementation:

```typescript
function Logger(...p: any) {
    console.log(...p);
    return true;
}
``` 

If you then want to use it as a handler (`whatsapp.on.message(Logger)`), TypeScript will throw an error because the return type isn't void.

Why `unknown` and not `any`? Because eslint hates `any` 😂

---

The Promise code you sent is the same as calling an async function. I think making it a method will reduce code duplication, as it will be used for `on.message`, `on.status` and `on.sent`. The try catch might be problematic because it interrupts the stacktrace (Node throws a warning if a promise fails and isn't handled, so there's no need to log our own neither). However, I think we could test both handled and unhandled calls and check which one results in a better dev experience.

**Comment by RahulLanjewar93** [comment]
I've added the fix for emitters, 
Let's continue the discussion on the promisify thing on a sperate thread.
